panic: assertion failed: Page expected to be: 5

### Bug report criteria

- [ ] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [ ] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [ ] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [ ] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

Panic discovered by Antithesis testing

https://linuxfoundation.antithesis.com/report/G-9rIjiZJiwodTEN5avQ7wgK/u_uFsWOwZSxS5mOmbEprwMUijNhsWdV6mfde_CT-y4k.html?auth=v2.public.eyJuYmYiOiIyMDI1LTA3LTAzVDA3OjMxOjUyLjQzMzQ2ODk3NFoiLCJzY29wZSI6eyJSZXBvcnRTY29wZVYxIjp7ImFzc2V0IjoidV91RnNXT3daU3hTNW1PbWJFcHJ3TVVpak5oc1dkVjZtZmRlX0NULXk0ay5odG1sIiwicmVwb3J0X2lkIjoiRy05cklqaVpKaXdvZFRFTjVhdlE3d2dLIn19fUHU0wnVRoDtfilwOCROUiDTtcOlIZkrVaddCqjorH3utgcIEPIzlsrMAJGXFC6NTZMneLqAWWU_lq-9prD_tQc

/cc @ahrtr @fuweid 

### What did you expect to happen?

No panic

### How can we reproduce it (as minimally and precisely as possible)?

Contact @marcus-hodgson-antithesis to help reproduce it on Antithesis 

### Anything else we need to know?

_No response_

### Etcd version (please run commands below)

main branch

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>


### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>


### Relevant log output

```Shell
{"level":"info","ts":"2025-07-02T17:53:12.524581Z","caller":"snap/db.go:65","msg":"saved database snapshot to disk","path":"/var/etcd/data/member/snap/0000000000000051.snap.db","bytes":24576,"size":"25 kB"}
{"level":"info","ts":"2025-07-02T17:53:12.524643Z","caller":"rafthttp/http.go:287","msg":"received and saved database snapshot","local-member-id":"dcb68c82481661be","remote-snapshot-sender-id":"99eab3685d8363a1","incoming-snapshot-index":81,"incoming-snapshot-size-bytes":24576,"incoming-snapshot-size":"25 kB","download-took":"1.602245032s"}
{"level":"info","ts":"2025-07-02T17:53:12.525182Z","logger":"raft","caller":"v3@v3.6.0/raft.go:897","msg":"dcb68c82481661be became follower at term 147"}
{"level":"info","ts":"2025-07-02T17:53:12.525209Z","logger":"raft","caller":"v3@v3.6.0/log.go:465","msg":"log [committed=79, applied=79, applying=79, unstable.offset=80, unstable.offsetInProgress=80, len(unstable.Entries)=0] starts to restore snapshot [index: 81, term: 147]"}
{"level":"info","ts":"2025-07-02T17:53:12.525245Z","logger":"raft","caller":"v3@v3.6.0/raft.go:1981","msg":"dcb68c82481661be switched to configuration voters=(3201069280892892151 11090874293180130209 15904053625288548798)"}
{"level":"info","ts":"2025-07-02T17:53:12.525340Z","logger":"raft","caller":"v3@v3.6.0/raft.go:1935","msg":"dcb68c82481661be [commit: 81, lastindex: 81, lastterm: 147] restored snapshot [index: 81, term: 147]"}
{"level":"info","ts":"2025-07-02T17:53:12.525381Z","logger":"raft","caller":"v3@v3.6.0/raft.go:1844","msg":"dcb68c82481661be [commit: 81] restored snapshot [index: 81, term: 147]"}
{"level":"info","ts":"2025-07-02T17:53:12.525408Z","logger":"raft","caller":"v3@v3.6.0/node.go:370","msg":"raft.node: dcb68c82481661be elected leader 99eab3685d8363a1 at term 147"}
{"level":"info","ts":"2025-07-02T17:53:12.525980Z","caller":"etcdserver/raft.go:276","msg":"applied incoming Raft snapshot","snapshot-index":81}
{"level":"info","ts":"2025-07-02T17:53:12.526036Z","caller":"etcdserver/server.go:1001","msg":"applying snapshot","current-snapshot-index":52,"current-applied-index":79,"incoming-leader-snapshot-index":81,"incoming-leader-snapshot-term":147}
{"level":"info","ts":"2025-07-02T17:53:12.526095Z","logger":"bbolt","caller":"backend/backend.go:203","msg":"Opening db file (/var/etcd/data/member/snap/db) with mode -rw------- and with options: {Timeout: 0s, NoGrowSync: false, NoFreelistSync: true, PreLoadFreelist: false, FreelistType: hashmap, ReadOnly: false, MmapFlags: 8000, InitialMmapSize: 10737418240, PageSize: 0, NoSync: false, OpenFile: 0x0, Mlock: false, Logger: 0xc000020d00}"}
{"level":"info","ts":"2025-07-02T17:53:12.526699Z","logger":"bbolt","caller":"runtime/panic.go:792","msg":"Opening bbolt db (/var/etcd/data/member/snap/db) successfully"}
panic: assertion failed: Page expected to be: 5, but self identifies as 0
goroutine 8052 [running]:
go.etcd.io/bbolt/internal/common.Assert(...)
	go.etcd.io/bbolt@v1.4.2/internal/common/verify.go:65
go.etcd.io/bbolt/internal/common.(*Page).FastCheck(0x7f16ce802000, 0x5)
	go.etcd.io/bbolt@v1.4.2/internal/common/page.go:83 +0x1d9
go.etcd.io/bbolt.(*Tx).page(0xc0010a31d0?, 0xc0002992e8?)
	go.etcd.io/bbolt@v1.4.2/tx.go:598 +0x7b
go.etcd.io/bbolt.(*Tx).forEachPageInternal(0xc000c71340, {0xc0010a31d0, 0x1, 0xa}, 0xc0002993a0)
	go.etcd.io/bbolt@v1.4.2/tx.go:610 +0x5a
go.etcd.io/bbolt.(*Tx).forEachPage(...)
	go.etcd.io/bbolt@v1.4.2/tx.go:606
go.etcd.io/bbolt.(*Tx).checkInvariantProperties(0xc000c71340, 0x5, 0xc0001ee808?, 0xc00131e780?, {0x1526270, 0x1e93dc0}, 0xc001329570)
	go.etcd.io/bbolt@v1.4.2/tx_check.go:143 +0xb2
go.etcd.io/bbolt.(*Tx).recursivelyCheckBucket(0xc000c71340, 0xc000c71358, 0xc000299640, 0xc000299588, {0x1526270, 0x1e93dc0}, 0xc001329570)
	go.etcd.io/bbolt@v1.4.2/tx_check.go:130 +0x73
go.etcd.io/bbolt.(*DB).freepages(0xc00048cfc8)
	go.etcd.io/bbolt@v1.4.2/db.go:1251 +0x21b
go.etcd.io/bbolt.(*DB).loadFreelist.func1()
	go.etcd.io/bbolt@v1.4.2/db.go:422 +0xc9
sync.(*Once).doSlow(0xc000299708?, 0xc00048cfc8?)
	sync/once.go:78 +0xab
sync.(*Once).Do(...)
	sync/once.go:69
go.etcd.io/bbolt.(*DB).loadFreelist(0xc00048cfc8?)
	go.etcd.io/bbolt@v1.4.2/db.go:418 +0x3b
go.etcd.io/bbolt.Open({0xc0011f2600, 0x1d}, 0x180, 0xc00136a070)
	go.etcd.io/bbolt@v1.4.2/db.go:299 +0xb48
go.etcd.io/etcd/server/v3/storage/backend.newBackend({{0xc0011f2600, 0x1d}, 0x5f5e100, 0x2710, {0x12d2af4, 0x7}, 0x280000000, 0xc000213780, 0x0, 0x0, ...})
	server/storage/backend/backend.go:203 +0x21b
go.etcd.io/etcd/server/v3/storage/backend.New({{0xc0011f2600, 0x1d}, 0x5f5e100, 0x2710, {0x12d2af4, 0x7}, 0x280000000, 0xc000213780, 0x0, 0x0, ...})
	server/storage/backend/backend.go:167 +0x45
go.etcd.io/etcd/server/v3/storage.newBackend({{0xc00001203a, 0x5}, {{{0x1e93dc0, 0x0, 0x0}, 0x12a05f200, 0x77359400, 0x77359400, 0x165a0bc00, 0x0, ...}, ...}, ...}, ...)
	server/storage/backend.go:55 +0x478
go.etcd.io/etcd/server/v3/storage.OpenBackend.func1()
	server/storage/backend.go:76 +0x8b
created by go.etcd.io/etcd/server/v3/storage.OpenBackend in goroutine 171
	server/storage/backend.go:75 +0x16f
```

--- a/server/etcdserver/api/v3rpc/maintenance.go
+++ b/server/etcdserver/api/v3rpc/maintenance.go
@@ -41,6 +41,10 @@ type BackendGetter interface {
 	Backend() backend.Backend
 }
 
+type Defrager interface {
+	Defragment() error
+}
+
 type Alarmer interface {
 	// Alarms is implemented in Server interface located in etcdserver/server.go
 	// It returns a list of alarms present in the AlarmStore
@@ -71,6 +75,7 @@ type maintenanceServer struct {
 	hasher mvcc.HashStorage
 	kg     KVGetter
 	bg     BackendGetter
+	defrag Defrager
 	a      Alarmer
 	lt     LeaderTransferrer
 	hdr    header
@@ -81,7 +86,7 @@ type maintenanceServer struct {
 }
 
 func NewMaintenanceServer(s *etcdserver.EtcdServer, healthNotifier notifier) pb.MaintenanceServer {
-	srv := &maintenanceServer{lg: s.Cfg.Logger, rg: s, hasher: s.KV().HashStorage(), kg: s, bg: s, a: s, lt: s, hdr: newHeader(s), cs: s, d: s, healthNotifier: healthNotifier}
+	srv := &maintenanceServer{lg: s.Cfg.Logger, rg: s, hasher: s.KV().HashStorage(), kg: s, bg: s, defrag: s, a: s, lt: s, hdr: newHeader(s), cs: s, d: s, healthNotifier: healthNotifier}
 	if srv.lg == nil {
 		srv.lg = zap.NewNop()
 	}
@@ -92,7 +97,7 @@ func (ms *maintenanceServer) Defragment(ctx context.Context, sr *pb.DefragmentRe
 	ms.lg.Info("starting defragment")
 	ms.healthNotifier.defragStarted()
 	defer ms.healthNotifier.defragFinished()
-	err := ms.bg.Backend().Defrag()
+	err := ms.defrag.Defragment()
 	if err != nil {
 		ms.lg.Warn("failed to defragment", zap.Error(err))
 		return nil, err

--- a/server/etcdserver/server.go
+++ b/server/etcdserver/server.go
@@ -1334,11 +1334,21 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 	// wait for raftNode to persist snapshot onto the disk
 	<-apply.notifyc
 
+	bemuUnlocked := false
+	s.bemu.Lock()
+	defer func() {
+		if !bemuUnlocked {
+			s.bemu.Unlock()
+		}
+	}()
+
 	// gofail: var applyBeforeOpenSnapshot struct{}
 	newbe, err := openSnapshotBackend(s.Cfg, s.snapshotter, apply.snapshot, s.beHooks)
 	if err != nil {
 		lg.Panic("failed to open snapshot backend", zap.Error(err))
 	}
+	lg.Info("applySnapshot: opened snapshot backend")
+	// gofail: var applyAfterOpenSnapshot struct{}
 
 	// We need to set the backend to consistIndex before recovering the lessor,
 	// because lessor.Recover will commit the boltDB transaction, accordingly it
@@ -1366,11 +1376,14 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 	newbe.SetTxPostLockInsideApplyHook(s.getTxPostLockInsideApplyHook())
 	lg.Info("restored mvcc store", zap.Uint64("consistent-index", s.consistIndex.ConsistentIndex()))
 
+	oldbe := s.be
+	s.be = newbe
+	s.bemu.Unlock()
+	bemuUnlocked = true
+
 	// Closing old backend might block until all the txns
 	// on the backend are finished.
 	// We do not want to wait on closing the old backend.
-	s.bemu.Lock()
-	oldbe := s.be
 	go func() {
 		lg.Info("closing old backend file")
 		defer func() {
@@ -1380,9 +1393,6 @@ func (s *EtcdServer) applySnapshot(ep *etcdProgress, apply *apply) {
 			lg.Panic("failed to close old backend", zap.Error(err))
 		}
 	}()
-
-	s.be = newbe
-	s.bemu.Unlock()
 
 	lg.Info("restoring alarm store")
 
@@ -2930,4 +2940,10 @@ func maybeDefragBackend(cfg config.ServerConfig, be backend.Backend) error {
 
 func (s *EtcdServer) CorruptionChecker() CorruptionChecker {
 	return s.corruptionChecker
+}
+
+func (s *EtcdServer) Defragment() error {
+	s.bemu.Lock()
+	defer s.bemu.Unlock()
+	return s.be.Defrag()
 }

--- a/tests/e2e/reproduce_20271_test.go
+++ b/tests/e2e/reproduce_20271_test.go
@@ -1,0 +1,107 @@
+// Copyright 2025 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+// TestIssue20271 reproduces the issue: https://github.com/etcd-io/etcd/issues/20271.
+func TestIssue20271(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	ctx := t.Context()
+
+	snapCount := 10
+
+	epc, err := e2e.NewEtcdProcessCluster(t,
+		&e2e.EtcdProcessClusterConfig{
+			SnapshotCount:          snapCount,
+			SnapshotCatchUpEntries: snapCount,
+			ClusterSize:            3,
+			KeepDataDir:            true,
+			GoFailEnabled:          true,
+			LogLevel:               "debug",
+			RollingStart:           true,
+		})
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, epc.Close())
+	}()
+
+	t.Log("Step 0: Ensure the leader is not the third member")
+	if leaderIdx := epc.WaitLeader(t); leaderIdx == 2 {
+		var transferee uint64
+		firstCli := newClient(t, epc.Procs[0].EndpointsGRPC(), e2e.ClientNonTLS, false)
+		statusResp, err := firstCli.Status(ctx, epc.Procs[0].EndpointsV3()[0])
+		require.NoError(t, err)
+		transferee = statusResp.Header.GetMemberId()
+
+		cli := newClient(t, epc.Procs[leaderIdx].EndpointsGRPC(), e2e.ClientNonTLS, false)
+		_, err = cli.MoveLeader(ctx, transferee)
+		require.NoError(t, err)
+	}
+
+	t.Log("Step 1: Write some data to the cluster")
+	for i := 0; i < snapCount*5; i++ {
+		require.NoError(t, epc.Procs[0].Etcdctl(e2e.ClientNonTLS, false, false).
+			Put(fmt.Sprintf("foo%d", i), strings.Repeat("Oops", 1024)))
+	}
+
+	t.Log(`Step 2: Config the third member to sleep 15s after OpenSnapshotBackend and use SIGSTOP to pause it.`)
+	require.NoError(t, epc.Procs[2].Failpoints().SetupHTTP(ctx, "applyAfterOpenSnapshot", `sleep("15s")`))
+	epc.Procs[2].Pause()
+
+	t.Log("Step 3: Write some key values to trigger new snapshot on the first two members")
+	for i := 0; i < snapCount+20; i++ {
+		err = epc.Procs[0].Etcdctl(e2e.ClientNonTLS, false, false).Put(fmt.Sprintf("foo%d", i), strings.Repeat("Awoo", 10))
+		require.NoError(t, err)
+	}
+
+	t.Log("Step 4: Restarting the first two members to re-connect to the paused member, so the inflight messages will be dropped. This will trigger new leader to send snapshot file to the third member.")
+	for _, proc := range epc.Procs[:2] {
+		require.NoError(t, proc.Restart())
+	}
+	epc.Procs[2].Resume()
+
+	t.Log(`Step 5: After opening snapshot file from new leader, invoke defragment\n
+to override boltdb file. So, for the following changes, the third member will commit them into deleted boltdb file.`)
+	e2e.AssertProcessLogs(t, epc.Procs[2], "applySnapshot: opened snapshot backend")
+	err = epc.Procs[2].Etcdctl(e2e.ClientNonTLS, false, false).Defragment(20 * time.Second)
+	require.NoError(t, err)
+	e2e.AssertProcessLogs(t, epc.Procs[2], "applied snapshot")
+	require.NoError(t, epc.Procs[2].Failpoints().DeactivateHTTP(ctx, "applyAfterOpenSnapshot"))
+
+	t.Log("Step 6: Write some data to the cluster")
+	for i := 0; i < snapCount/2; i++ {
+		err = epc.Procs[0].Etcdctl(e2e.ClientNonTLS, false, false).Put(fmt.Sprintf("foo%d", i), strings.Repeat("Oops", 1))
+		require.NoError(t, err)
+	}
+
+	t.Log("Step 7: Restart the third member. It recovers from the new boltdb file. Therefore, data writen in Step 6 is lost.")
+	require.NoError(t, epc.Procs[2].Restart())
+
+	t.Log("Step 8: Check hashkv of each member")
+	assertKVHash(t, epc)
+}

--- a/tests/framework/e2e/etcd_process.go
+++ b/tests/framework/e2e/etcd_process.go
@@ -56,6 +56,9 @@ type EtcdProcess interface {
 	Stop() error
 	Close() error
 	Kill() error
+	Pause() error
+	Resume() error
+
 	WithStopSignal(sig os.Signal) os.Signal
 	Config() *EtcdServerProcessConfig
 	Logs() LogsExpect
@@ -222,6 +225,16 @@ func (ep *EtcdServerProcess) Close() error {
 func (ep *EtcdServerProcess) Kill() error {
 	ep.cfg.lg.Info("killing server...", zap.String("name", ep.cfg.Name))
 	return ep.proc.Signal(syscall.SIGKILL)
+}
+
+func (ep *EtcdServerProcess) Pause() error {
+	ep.cfg.lg.Info("Pausing server...", zap.String("name", ep.cfg.Name))
+	return ep.proc.Signal(syscall.SIGSTOP)
+}
+
+func (ep *EtcdServerProcess) Resume() error {
+	ep.cfg.lg.Info("Resuming server...", zap.String("name", ep.cfg.Name))
+	return ep.proc.Signal(syscall.SIGCONT)
 }
 
 func (ep *EtcdServerProcess) Wait(ctx context.Context) error {

